products/libllbuild/BuildSystem-C-API.cpp

```diff
@@ -292,6 +293,58 @@ class CAPIBuildSystemFrontendDelegate : public BuildSystemFrontendDelegate {
   void cancel() override {
     BuildSystemFrontendDelegate::cancel();
   }
+
+  virtual void cycleDetected(const std::vector<core::Rule*>& items) override {
+    llb_build_key_t* rules = (llb_build_key_t*)malloc(items.size() * sizeof(llb_build_key_t));
```

Comment:
It would be safer to use `std::vector<llb_build_key>(<N>)` here, and just pass its pointer to the client. That avoids the need to manage the memory for this particular thing ourselves.